### PR TITLE
Notebooks

### PR DESCRIPTION
**Pull request by johnbensnyder**
Adding in 3 notebooks for users
1) coco inspection gives details of the coco dataset and pycocotools, and provides scripts to create smaller subsets of the data
2) tensor inspection provides instructions for outputing tensors as the model is running
3) visualization provides tools to apply a model to images and observe the predictions

Also added in new data tools to support functionality in these notebooks, as well as a pretrained model stored under releases.

### CODE REVIEW

File: MaskRCNN/train.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 # SPDX-License-Identifier: Apache-2.0
-#!/usr/bin/env python
+# !/usr/bin/env python
```
Comment (YangFei1990) [inline MaskRCNN/train.py]: Is this a typo?

File: MaskRCNN/train.py [hunk 2]
```diff
@@ -62,13 +63,33 @@ def do_visualize(model, model_path, nr_visualize=100, output_dir='output'):
...
+            #    new_boxes.append(box)
+            #rpn_boxes = np.array(new_boxes)
+            #np.save('/data/final_boxes', final_boxes)
+            #new_boxes = []
+            #for box in final_boxes:
+            #    if box[2]>img.shape[1]:
+            #        box[2]=img.shape[1]
+            #    new_boxes.append(box)
+            #final_boxes = np.array(new_boxes)
+            #final_boxes = np.array([i[[1,0,3,2]] for i in final_boxes])
```
Comment (YangFei1990) [inline MaskRCNN/train.py]: Do you need those comments? If no, please clear.

File: MaskRCNN/data.py [hunk 3]
```diff
@@ -478,6 +478,115 @@ def preprocess(roidb):
...
+    anchor_labels: (h', w', NA)
+    anchor_boxes: (h', w', NA, 4)
+
+    gt_boxes: (N, 4)
+    gt_labels: (N,)
+
+    If MODE_MASK, gt_masks: (N, h, w)
+    """
+
+    roidbs = DetectionDataset().load_viz_roidbs(name)
```
Comment (YangFei1990) [inline MaskRCNN/data.py]: If the only difference between `get_viz_dataflow` and `get_train_dataflow` is the loading roidb part, we better merge these two functions together.

**Comment by YangFei1990** [comment]
Do we also consider to add jupyter to our container? Currently we need to manually do `pip install jupyter` inside container, correct me if I'm wrong.

File: notebooks/README.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
+## Launching notebooks
+
+These notebooks are designed to be used within the mask r-cnn docker container on a P3 EC2 instance. After starting an instance, get the public IP address, and ssh from your terminal using
+
+```ssh ~/your/pem/file localhost:8890:localhost:8888 ubuntu@[ip-address]```
```
Comment (YangFei1990) [inline notebooks/README.md]: Did you miss the -i and -L option?

File: notebooks/README.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+This will ssh into your instance, while forwarding whatever is running on ports 8888 and 6006 in your instance to ports 8890 and 6008 on your local machine.
```
Comment (YangFei1990) [inline notebooks/README.md]: > running on ports 8888 and 6006 in your instance to ports 8890 and 6008 

Do we really have the 6006&6008 port?

File: notebooks/README.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+```
+Currently running servers:
+http://0.0.0.0:8888/?token=d4414589af37ed69b1efce555c8eddc3139e5b3baaa4ff05 :: /
+```
+
+Copy the token portion ```d4414589af37ed69b1efce555c8eddc3139e5b3baaa4ff05``` and go to ```localhost:8890``` in your browser.
+
+From the Jupyter homepage, you can browse files within the container, upload and download files, create a new notebook, or start a terminal from within the container.
+```
```
Comment (YangFei1990) [inline notebooks/README.md]: Can you give a brief description of each notebook, something like:
For each note book:
- A brief intro
  - functionality 1
  - functionality 2
  ...

File: Dockerfile [hunk 7]
```diff
@@ -11,7 +11,12 @@ RUN apt-get update && \
     pip install opencv-python
 
 RUN pip uninstall -y pycocotools && \
-    pip install pybind11
+    pip install pybind11 && \
+    pip install scikit-image
+    
+RUN wget https://github.com/aws-samples/mask-rcnn-tensorflow/releases/download/v0.0.0/example_log.tar.gz && \
+    tar -xzf example_log.tar.gz example_log
```
Comment (YangFei1990) [inline Dockerfile]: What is this for? Are you sure it can run? When I ran it in the container I encountered:
```
tar: example_log: Not found in archive
```

File: notebooks/tensor_inspection.ipynb [hunk 8]
```diff
@@ -0,0 +1,1028 @@
+{
```
Comment (pidajay) [inline notebooks/tensor_inspection.ipynb]: General comment on the readability aspect of this notebook. The formatting is a bit all over the place.

File: notebooks/visualization.ipynb [hunk 9]
```diff
@@ -0,0 +1,424 @@
+{
```
Comment (pidajay) [inline notebooks/visualization.ipynb]: It would be nice to have the cells executed for this notebook

File: MaskRCNN/data.py (hunk 3 above)
Comment (johnbensnyder) [reply MaskRCNN/data.py]: Ideally, I'd like to merge these two and the eval dataflow, since they all only have minor differences. I'd rather do that in a separate pull request though, because I don't want to risk breaking anything in the get_train_dataflow.

File: Dockerfile (hunk 7 above)
Comment (johnbensnyder) [reply Dockerfile]: This downloads a pretrained model for visualization. Will do some testing to see if it needs to be updated.

File: notebooks/visualization.ipynb (hunk 9 above)
Comment (johnbensnyder) [reply notebooks/visualization.ipynb]: Uploaded a new version with come cells executed. Executing all of them makes the notebook too large to use in the browser.

File: notebooks/tensor_inspection.ipynb (hunk 8 above)
Comment (johnbensnyder) [reply notebooks/tensor_inspection.ipynb]: This is an issue with Github, not sure if there's a good way around it, but the notebook looks better when used outside github.

File: MaskRCNN/data.py (hunk 3 above)
Comment (YangFei1990) [reply MaskRCNN/data.py]: The eval data flow is quite different but we do can merge the get_viz_dataflow and get_train_dataflow. 
You are right we can have a different PR for that.

File: Dockerfile (hunk 7 above)
Comment (johnbensnyder) [reply Dockerfile]: Problem was an error in the tar file itself. Uploaded a new one. Can you try again?
